File: IdentityX/testing_framework/project_test.go
package testing

import (
	"GoAuth/internal/errx"
	"encoding/base64"
	"net/http"
	"testing"

	"github.com/stretchr/testify/require"
)

func testProjects(t *testing.T, suite *TestSuite) {
	client := suite.NewClient(t)
	user := client.WithCredentials("projects@mail.com", ValidPassword).Register().Login()

	var projectID string
	t.Run("CreateProject", func(t *testing.T) {
		authClient := suite.NewClient(t).WithAuth(user.auth)
		val := authClient.POST("/projects").
			WithBody(map[string]interface{}{
				"project_name": "Test Project",
				"metadata":     map[string]string{"env": "test"},
			}).
			Expect(http.StatusCreated).
			HasMessage("Created project").
			RequireDataValue()

		spec := map[string]interface{}{
			"id":           StoreString{Into: &projectID, Matcher: AnyUUID{}},
			"owner_id":     AnyUUID{},
			"project_name": "Test Project",
			"is_active":    true,
			"metadata": map[string]interface{}{
				"env": "test",
			},
			"created_at": AnyDate{},
			"updated_at": AnyDate{},
		}

		Validate(t, val, spec)
	})

	t.Run("ValidationCreateProject", func(t *testing.T) {
		authClient := suite.NewClient(t).WithAuth(user.auth)
		authClient.POST("/projects").
			WithBody(map[string]interface{}{
				"project_name": "", // Empty name
				"metadata":     map[string]string{"env": "test"},
			}).
			Expect(http.StatusBadRequest).
			HasErrID(errx.RequestValidationError).
			ValidationError("project_name is required")
	})

	t.Run("ListProjects", func(t *testing.T) {
		authClient := suite.NewClient(t).WithAuth(user.auth)
		data := authClient.GET("/projects").
			Expect(http.StatusOK).
			RequireDataValue()

		spec := []interface{}{
			map[string]interface{}{
				"id":           AsString{projectID, AnyUUID{}},
				"owner_id":     AnyUUID{},
				"project_name": "Test Project",
				"is_active":    true,
				"metadata": map[string]interface{}{
					"env": "test",
				},
				"created_at": AnyDate{},
				"updated_at": AnyDate{},
			},
		}

		Validate(t, data, spec)
	})

	t.Run("GetProject", func(t *testing.T) {
		authClient := suite.NewClient(t).WithAuth(user.auth)
		data := authClient.GET("/projects/" + projectID).
			Expect(http.StatusOK).
			RequireDataValue()

		spec := map[string]interface{}{
			"id":           AsString{projectID, AnyUUID{}},
			"owner_id":     AnyUUID{},
			"project_name": "Test Project",
			"is_active":    true,
			"metadata": map[string]interface{}{
				"env": "test",
			},
			"created_at": AnyDate{},
			"updated_at": AnyDate{},
		}

		Validate(t, data, spec)
	})

	t.Run("ListProjectUsers", func(t *testing.T) {
		authClient := suite.NewClient(t).WithAuth(user.auth)

		// Register a user to this project first
		projectUserEmail := "projectuser@mail.com"
		suite.NewClient(t).POST("/projects/" + projectID + "/register").
			WithBody(map[string]interface{}{
				"email":    projectUserEmail,
				"password": ValidPassword,
			}).
			Expect(http.StatusCreated)

		data := authClient.GET("/projects/" + projectID + "/users").
			Expect(http.StatusOK).
			RequireDataValue()

		spec := []interface{}{
			map[string]interface{}{
				"id":         AnyUUID{},
				"project_id": AsString{projectID, AnyUUID{}},
				"email":      projectUserEmail,
				"user_type":  "project",
				"is_active":  true,
				"created_at": AnyDate{},
				"updated_at": AnyDate{},
			},
		}

		Validate(t, data, spec)
	})

	t.Run("CrossUserAccess", func(t *testing.T) {
		// Create a second user
		attacker := client.WithCredentials("attacker@mail.com", ValidPassword).Register().Login()
		attackerClient := suite.NewClient(t).WithAuth(attacker.auth)

		// Try to GET project owned by first user
		attackerClient.GET("/projects/" + projectID).
			Expect(http.StatusNotFound).
			HasErrID(errx.SQLNotFound).
			HasMessage("project not found")

		// Try to UPDATE
		attackerClient.PATCH("/projects/" + projectID).
			WithBody(map[string]interface{}{
				"project_name": "Hacked",
			}).
			Expect(http.StatusNotFound).
			HasErrID(errx.SQLNotFound).
			HasMessage("project not found")

		// Try to DELETE
		attackerClient.DELETE("/projects/" + projectID).
			Expect(http.StatusNotFound).
			HasErrID(errx.ProjectNotFound).
			HasMessage("project not found")

		// Ensure it was NOT actually deleted from the perspective of the owner
		authClient := suite.NewClient(t).WithAuth(user.auth)
		authClient.GET("/projects/" + projectID).
			Expect(http.StatusOK)
	})

	t.Run("UpdateProject", func(t *testing.T) {
		authClient := suite.NewClient(t).WithAuth(user.auth)
		data := authClient.PATCH("/projects/" + projectID).
			WithBody(map[string]interface{}{
				"project_name": "Updated Project",
				"metadata":     map[string]interface{}{"env": "prod"},
			}).
			Expect(http.StatusOK).
			RequireDataValue()

		spec := map[string]interface{}{
			"id":           AsString{projectID, AnyUUID{}},
			"owner_id":     AnyUUID{},
			"project_name": "Updated Project",
			"is_active":    true,
			"metadata": map[string]interface{}{
				"env": "prod",
			},
			"created_at": AnyDate{},
			"updated_at": AnyDate{},
		}

		Validate(t, data, spec)
	})

	t.Run("GetProjectJWKS", func(t *testing.T) {
		authClient := suite.NewClient(t).WithAuth(user.auth)
		obj := authClient.GET("/projects/" + projectID + "/.well-known/jwks.json").
			Expect(http.StatusOK).
			JSONObj()

		obj.Value("keys").Array().NotEmpty()

		// Verify decoding (Client gets it and decodes)
		xBase64 := obj.Value("keys").Array().Value(0).Object().Value("x").String().Raw()
		xBytes, err := base64.RawURLEncoding.DecodeString(xBase64)
		require.NoError(t, err)
		pubKey := parseJWKXToEd25519PublicKey(t, xBytes)
		require.NotNil(t, pubKey)

		// Unauthenticated access denial
		client.GET("/projects/" + projectID + "/.well-known/jwks.json").
			Expect(http.StatusUnauthorized)
	})

	t.Run("DeleteProject", func(t *testing.T) {
		authClient := suite.NewClient(t).WithAuth(user.auth)
		authClient.DELETE("/projects/" + projectID).
			Expect(http.StatusOK).
			HasMessage("Deleted project")

		// Verify deletion
		authClient.GET("/projects").
			Expect(http.StatusOK).
			RequireDataArray().Length().IsEqual(0)
	})
}

File: IdentityX/testing_framework/project.test.ts
import { beforeAll, describe, expect, test } from "vitest";
import { createClient } from "./helpers/index.js";
import { assertErrID, assertMessage, assertValidationError, shouldFail } from "./helpers/assert.js";
import { Validate, AnyUUID, AnyDate, AsString } from "./helpers/validate.js";
import { ValidPassword } from "./fixtures/auth/testdata.js";

// Error IDs
// RequestValidationError = fail.ID(0, "REQ", 4, false, ...) → 0_REQ_0004_D
// SQLNotFound            = fail.ID(0, "SQL", 0, false, ...) → 0_SQL_0000_D
// ProjectNotFound        = fail.ID(0, "PROJECT", 0, true, ...) → 0_PROJECT_0000_S
const ErrRequestValidation = "0_REQ_0004_D";
const ErrSQLNotFound       = "0_SQL_0000_D";
const ErrProjectNotFound   = "0_PROJECT_0000_S";

// ============================================================================
// Ed25519 JWK verification
// ============================================================================

function verifyEd25519JWKKey(x: string): void {
    // x is base64url-encoded raw public key bytes (32 bytes for Ed25519)
    const base64 = x.replace(/-/g, "+").replace(/_/g, "/");
    const padded = base64.padEnd(base64.length + (4 - (base64.length % 4)) % 4, "=");
    const bytes  = Buffer.from(padded, "base64");

    expect(bytes.length).toBe(32);
}

// ============================================================================
// PROJECTS TESTS
// ============================================================================

describe("Projects", () => {
    let user: Awaited<ReturnType<typeof createClient.prototype.login>>;
    let projectID: string;

    beforeAll(async () => {
        await createClient().withCredentials("projects@mail.com", ValidPassword).register();
        user = await createClient().withCredentials("projects@mail.com", ValidPassword).login();
    });

    test("CreateProject", async () => {
        const data = await user.post("/projects", {
            project_name: "Test Project",
            metadata: { env: "test" },
        });

        Validate(data, {
            id:           AnyUUID,
            owner_id:     AnyUUID,
            project_name: "Test Project",
            is_active:    true,
            metadata:     { env: "test" },
            created_at:   AnyDate,
            updated_at:   AnyDate,
        });

        projectID = data.id;
    });

    test("ValidationCreateProject", async () => {
        const body = await shouldFail(
            user.post("/projects", { project_name: "", metadata: { env: "test" } }),
            400
        );
        assertErrID(body, ErrRequestValidation);
        assertValidationError(body, "project_name is required");
    });

    test("ListProjects", async () => {
        const data = await user.get("/projects");

        Validate(data, [
            {
                id:           AsString(projectID, AnyUUID),
                owner_id:     AnyUUID,
                project_name: "Test Project",
                is_active:    true,
                metadata:     { env: "test" },
                created_at:   AnyDate,
                updated_at:   AnyDate,
            },
        ]);
    });

    test("GetProject", async () => {
        const data = await user.get(`/projects/${projectID}`);

        Validate(data, {
            id:           AsString(projectID, AnyUUID),
            owner_id:     AnyUUID,
            project_name: "Test Project",
            is_active:    true,
            metadata:     { env: "test" },
            created_at:   AnyDate,
            updated_at:   AnyDate,
        });
    });

    test("ListProjectUsers", async () => {
        const projectUserEmail = "projectuser@mail.com";

        await createClient().http.post(`/projects/${projectID}/register`, {
            email:    projectUserEmail,
            password: ValidPassword,
        });

        const data = await user.get(`/projects/${projectID}/users`);

        Validate(data, [
            {
                id:         AnyUUID,
                project_id: AsString(projectID, AnyUUID),
                email:      projectUserEmail,
                user_type:  "project",
                is_active:  true,
                created_at: AnyDate,
                updated_at: AnyDate,
            },
        ]);
    });

    test("CrossUserAccess", async () => {
        await createClient().withCredentials("attacker@mail.com", ValidPassword).register();
        const attacker = await createClient().withCredentials("attacker@mail.com", ValidPassword).login();

        // Try to GET
        const getBody = await shouldFail(attacker.get(`/projects/${projectID}`), 404);
        assertErrID(getBody, ErrSQLNotFound);
        assertMessage(getBody, "project not found");

        // Try to PATCH
        const patchBody = await shouldFail(
            attacker.patch(`/projects/${projectID}`, { project_name: "Hacked" }),
            404
        );
        assertErrID(patchBody, ErrSQLNotFound);
        assertMessage(patchBody, "project not found");

        // Try to DELETE
        const deleteBody = await shouldFail(attacker.del(`/projects/${projectID}`), 404);
        assertErrID(deleteBody, ErrProjectNotFound);
        assertMessage(deleteBody, "project not found");

        // Ensure owner can still access it
        const data = await user.get(`/projects/${projectID}`);
        expect(data.id).toBe(projectID);
    });

    test("UpdateProject", async () => {
        const data = await user.patch(`/projects/${projectID}`, {
            project_name: "Updated Project",
            metadata:     { env: "prod" },
        });

        Validate(data, {
            id:           AsString(projectID, AnyUUID),
            owner_id:     AnyUUID,
            project_name: "Updated Project",
            is_active:    true,
            metadata:     { env: "prod" },
            created_at:   AnyDate,
            updated_at:   AnyDate,
        });
    });

    test("GetProjectJWKS", async () => {
        const res = await user.http.get(`/projects/${projectID}/.well-known/jwks.json`, {
            headers: { Cookie: `access_token=${user.auth!.accessToken}; refresh_token=${user.auth!.refreshToken}` },
        });

        const keys: any[] = res.data.keys;
        expect(keys.length).toBeGreaterThan(0);

        // Verify the first key is a valid Ed25519 public key
        const x = keys[0].x as string;
        expect(typeof x).toBe("string");
        expect(x.length).toBeGreaterThan(0);
        verifyEd25519JWKKey(x);

        // Unauthenticated access should be denied
        const body = await shouldFail(
            createClient().http.get(`/projects/${projectID}/.well-known/jwks.json`),
            401
        );
        expect(body).toBeDefined();
    });

    test("DeleteProject", async () => {
        await user.del(`/projects/${projectID}`);

        const remaining = await user.get("/projects");
        expect(remaining).toHaveLength(0);
    });
});